extensions/src/main/java/com/tomtom/kotlin/extensions/CastExtensions.kt

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (C) 2012-2022, TomTom (http://tomtom.com).
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.tomtom.kotlin.extensions
+
+/**
+ * Casts `this` to `T`, throwing a [ClassCastException] if the cast fails.
+ *
+ * This can be used to replace:
+ *
+ * ```kotlin
+ * (
+ *     myObject
+ *         .chainCall()
+ *         .chainCall()
+ *         as MyClass
+ *  ).chainCall()
+ * ```
+ *
+ * or:
+ *
+ * ```kotlin
+ * myObject
+ *     .chainCall()
+ *     .chainCall()
+ *     .let { it as MyClass }
+ *     .chainCall()
+ * ```
+ *
+ * with a more functional-programming-friendly syntax:
+ *
+ * ```kotlin
+ * myObject
+ *     .chainCall()
+ *     .chainCall()
+ *     .cast<MyClass>()
+ *     .chainCall()
+ * ```
+ *
+ * This method is not intended to replace all casts; the `as` operator is fine in many cases. In
+ * chain-calling constructs however, the `as` operator can easily lead to reduced readability. `let`
+ * helps with this, but adds boilerplate that distracts from the functional flow.
+ */
+inline fun <reified T> Any.cast(): T =
```

Comment:
It works as a regular cast, meaning it will succeed because the generic type is erased. It doesn't show an unchecked cast warning however. It turns out the warning was not shown here either because of the `reified` (which is a leftover from an earlier version that I'll remove). Unfortunately the `inline` keyword doesn't also inline the warning.

To mitigate this, along with explicit documentation about it, I can also add an IDEA inspection for it. It's trivial to set up and will flag unchecked casts as warning/error. There is unfortunately no option to publish inspections for libraries by default (that I'm aware of), but for our own projects we can include it to make sure it's the default for everyone. Our current CI jobs do not take IDEA inspections into account, but Maxim added it to their backlog to change that. (I believe that's necessary anyway to address a large amount of other issues that currently go ignored.) It's definitely not as good as a compiler check, but it does go a long way.

All things considered I do believe this util is still worthwhile. Personally I experience very little value from the unchecked cast warnings - most often by far they're false positives. I find the added readability to outweigh the lack of warning (especially with the mentioned mitigation). I do realise it's debatable however.

As an alternative we can also consider a conditional `require(T::class.typeParameters.isEmpty())` inside the util. Disabling that `require` could be done through `cast<List<X>>(UNCHECKED_CAST)` (where `UNCHECKED_CAST` would be a constant of sorts).

Thoughts?

Reply:
(Rijn got a bit overexcited with the changes and merged the PR already, but I'll address this in a follow-up PR. 🙂)